File: front/go/deepx/attention.go
package deepx

import (
	"math"
)

// MultiHeadAttention 模块
type MultiHeadAttention struct {
	ModuleBase
	q_proj      *Linear
	k_proj      *Linear
	v_proj      *Linear
	o_proj      *Linear
	num_heads   int
	hidden_size int
}

func NewMultiHeadAttention(name string, hidden_size, num_heads int, dtype Dtype, g *Graph) *MultiHeadAttention {
	if g == nil {
		g = NewGraph()
	}
	if name == "" {
		name = "attention"
	}

	m := &MultiHeadAttention{
		ModuleBase: ModuleBase{
			g:    g,
			name: name,
		},
		q_proj:      NewLinear(name+".q_proj", hidden_size, hidden_size, dtype, g),
		k_proj:      NewLinear(name+".k_proj", hidden_size, hidden_size, dtype, g),
		v_proj:      NewLinear(name+".v_proj", hidden_size, hidden_size, dtype, g),
		o_proj:      NewLinear(name+".o_proj", hidden_size, hidden_size, dtype, g),
		num_heads:   num_heads,
		hidden_size: hidden_size,
	}
	return m
}

func (m *MultiHeadAttention) Forward(q, k, v *Tensor) *Tensor {
	batch_size := q.Shape.shape[0]
	seq_len := q.Shape.shape[1]
	head_dim := q.Shape.shape[2] / m.num_heads

	// 1. 线性变换
	query := m.q_proj.Linear(q)
	key := m.k_proj.Linear(k)
	value := m.v_proj.Linear(v)

	// 2. 重塑为多头形式
	// [batch_size, seq_len, hidden_size] -> [batch_size, seq_len, num_heads, head_dim]
	query = query.Reshape([]int{batch_size, seq_len, m.num_heads, head_dim})
	key = key.Reshape([]int{batch_size, seq_len, m.num_heads, head_dim})
	value = value.Reshape([]int{batch_size, seq_len, m.num_heads, head_dim})

	// 3. 转置以便进行批量矩阵乘法
	// [batch_size, seq_len, num_heads, head_dim] -> [batch_size, num_heads, seq_len, head_dim]
	query = query.Transpose([]int{0, 2, 1, 3})
	key = key.Transpose([]int{0, 2, 1, 3})
	value = value.Transpose([]int{0, 2, 1, 3})

	// 4. 计算注意力分数
	scores := query.Matmul(key)

	// 5. Scale
	d_k := float32(head_dim)
	scores = scores.Scale(1.0 / float32(math.Sqrt(float64(d_k))))

	// 6. Softmax
	attn := scores.Softmax()

	// 7. 加权求和
	out := attn.Matmul(value)

	// 8. 转置回原始形状
	// [batch_size, num_heads, seq_len, head_dim] -> [batch_size, seq_len, num_heads, head_dim]
	out = out.Transpose([]int{0, 2, 1, 3})

	// 9. 重塑回原始维度
	// [batch_size, seq_len, num_heads, head_dim] -> [batch_size, seq_len, hidden_size]
	out = out.Reshape([]int{batch_size, seq_len, m.hidden_size})

	// 10. 输出投影
	out = m.o_proj.Linear(out)

	return out
}

File: front/py/deepx/transformer/models/llama/attention.py
from typing import Optional,Tuple
from deepx.nn.modules import Module,Linear
from deepx import Tensor,matmul,softmax,concat,arange,dropout as dropout_func



def rotate_half(x:Tensor):
    index_front=arange(0,x.shape[-1]//2,dtype="int32")
    index_back=arange(x.shape[-1]//2,x.shape[-1],dtype="int32")
    x1 = x.indexselect(gatheraxis=-1,index=index_front)
    x2 = x.indexselect(gatheraxis=-1,index=index_back)
    return concat((-x2, x1,), dim=-1)

def apply_rotary_pos_emb(q:Tensor, k:Tensor, cos:Tensor, sin:Tensor, unsqueeze_dim:int=1):
    cos = cos.unsqueeze(unsqueeze_dim)
    sin = sin.unsqueeze(unsqueeze_dim)
    q_embed = (q * cos) + (rotate_half(q) * sin)
    k_embed = (k * cos) + (rotate_half(k) * sin)
    return q_embed, k_embed

def repeat_kv(hidden_states: Tensor, n_rep: int) -> Tensor:
    batch, num_key_value_heads, slen, head_dim = hidden_states.shape
    if n_rep == 1:
        return hidden_states
    hidden_states = hidden_states[:, :, None, :, :].expand(batch, num_key_value_heads, n_rep, slen, head_dim)
    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)


# https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py
# 经简化，去掉了分布式配置，去掉attention的配置。交给IR自动替换flashattention，后续的组件自动处理

def eager_attention_forward(
    module: Module,
    query: Tensor,
    key: Tensor,
    value: Tensor,
    attention_mask: Optional[Tensor],
    scaling: float,
    dropout: float = 0.0,
):
    key_states = repeat_kv(key, module.num_key_value_groups)
    value_states = repeat_kv(value, module.num_key_value_groups)

    attn_weights =  matmul(query, key_states.transpose(2, 3)) * scaling
    if attention_mask is not None:
        causal_mask = attention_mask[:, :, :, : key_states.shape[-2]]
        attn_weights = attn_weights + causal_mask

    attn_weights = softmax(attn_weights, dim=-1, dtype=query.dtype)
    attn_weights = dropout_func(attn_weights, p=dropout)
    attn_output = matmul(attn_weights, value_states)
    attn_output = attn_output.transpose(1, 2).contiguous()

    return attn_output, attn_weights

class LlamaAttention(Module):
    def __init__(self, config:dict, layer_idx: int):
        super().__init__()
        self.config = config
        self.layer_idx = layer_idx
        self.head_dim = getattr(config, "head_dim", config.hidden_size // config.num_attention_heads)
        self.num_key_value_groups = config.num_attention_heads // config.num_key_value_heads
        self.scaling = self.head_dim**-0.5
        self.attention_dropout = config.attention_dropout
        self.is_causal = True

        self.q_proj = Linear(
            config.hidden_size, config.num_attention_heads * self.head_dim, bias=config.attention_bias
        )
        self.k_proj = Linear(
            config.hidden_size, config.num_key_value_heads * self.head_dim, bias=config.attention_bias
        )
        self.v_proj = Linear(
            config.hidden_size, config.num_key_value_heads * self.head_dim, bias=config.attention_bias
        )
        self.o_proj = Linear(
            config.num_attention_heads * self.head_dim, config.hidden_size, bias=config.attention_bias
        )

    def forward(
        self,
        hidden_states:  Tensor,
        position_embeddings: Tuple[Tensor, Tensor],
        attention_mask: Optional[Tensor]
    ) -> Tuple[Tensor, Optional[Tensor], Optional[Tuple[Tensor]]]:
        input_shape = hidden_states.shape[:-1]
        hidden_shape = (*input_shape, -1, self.head_dim)

        query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)
        key_states = self.k_proj(hidden_states).view(hidden_shape).transpose(1, 2)
        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)

        cos, sin = position_embeddings
        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
 
 
        attn_output, attn_weights = attention_interface(
            self,
            query_states,
            key_states,
            value_states,
            attention_mask,
            dropout=0.0 if not self.training else self.attention_dropout,
            scaling=self.scaling,
            **kwargs,
        )

        attn_output = attn_output.reshape(*input_shape, -1)
        attn_output = self.o_proj(attn_output)
        return attn_output, attn_weights